### DISCUSSION TITLE
Understanding variable scope in Groovy

### DISCUSSION THREAD

**Discussion by timboxyz** [freeplane/freeplane#1955] [Help] [answered]
Because JS support is not in all JAVA 21 builds I am trying to convert my scripts to Groovy, which I am beginning to learn. In Groovy, by default according to all examples I have read, variable declarations are global similar to JavasScript.  However,  within the Freeplane environment variable declarations appear not to be global, instead, variables have to be explicitly passed into functions. This has caused some script failures. See a simplified example below that illustrates this:-

```
// testing scope of variables

def testtext= "some text"

//c.statusInfo = testtext // works

/* doesn't work
def display () {
    c.statusInfo = testtext
}
display()
*/

// works
def display (testtext) {
    c.statusInfo = testtext
}
display(testtext)
```
Before I commit to some substantial refactoring I am wondering if I have indeed surmised correctly or there is something else at play which would simplify the conversion?

My initial scripts were converted by one of the online  AI convertors which replaced all instances of `var `with `def` even though I understand` var` to be a legitimate alternative in Groovy. They did not however pick up the scope issue.

**Comment by i-plasm** [comment, marked as answer]
Hello! You can use the `@Field` annotation to increase the scope, this way :

~~~ groovy
//please include the `import ...` statement as well
import groovy.transform.Field

@Field def testtext= "some texts"

display()

def display () {
    c.statusInfo = testtext
}
~~~

**Reply by timboxyz** [reply]
> Hello! You can use the `@Field` annotation to increase the scope, this way :

Oh, TVM. So much for AI conversion, that is exactly the sort of thing I would expect it to know about.

It also completely ballsed up case-insensitive regexp which I had to fix.

If by default, scope is global, why do I need the extra code?

**Reply by i-plasm** [reply]
> If by default, scope is global, why do I need the extra code?

In your initial example, if you remove the `def` keyword from the variable, then it will be made "global" (in certain sense of the word, with dangerous caveats). I personally prefer the `def` or the `@Field def` declarations since I 99% times end up needing it, and 99% times would lead to error-prone-hard-to-debug code if I didn't do so -. [Here is an example](https://stackoverflow.com/questions/184002/groovy-whats-the-purpose-of-def-in-def-x-0/18026302#18026302) of an unexpected, hard-to-debug error. Anyways, your example would look like this with such approach:


~~~ groovy
// a "pseudo" global variable - notice nothing precedes the variable name
testtext= "some texts"

display()

def display () {
    c.statusInfo = testtext
}
~~~

**Reply by timboxyz** [reply]
Ahh, I think I see now. Because the script itself is run within some sort of class the definition is not being made at the top level of the whole code only wrt the script and so it's actually local. Makes sense.
